cpp/src/arrow/table.cc

```diff
@@ -612,7 +612,8 @@ Status TableBatchReader::ReadNext(std::shared_ptr<RecordBatch>* out) {
   }
 
   // Determine the minimum contiguous slice across all columns
-  int64_t chunksize = std::min(table_.num_rows(), max_chunksize_);
+  int64_t chunksize =
+      std::min(table_.num_rows() - absolute_row_position_, max_chunksize_);
```

Comment:
Ah. Is/would the other fix have been to change the `==` above to `>=`? (Though that would prevent the loop, but would give you a batch of the wrong size.)

Reply:
Basically, this adapts for the "loop remainder"
```c++
int64_t i = 0;
for (; i < table_.num_rows(); i += max_chunksize_) {
  process(max_chunksize_);
}
if (i < table_.num_rows()) {
  process(table_num_rows() - i);
}

Reply:
Yep, `>=` would avoid the infinite loop but not the incorrectly sized batch (which would be a problem).

Reply:
I'm not sure I follow on the "loop remainder" but what I have here works ok I'm pretty sure.

Reply:
Sorry, I am not saying to rewrite the logic as such, simply stating that conceptually this is what the code is doing now. You have a loop iterating with a constant non-unit stride but the total number of iterations is not a multiple of the stride, so at the end you are left with an iteration using a smaller stride. This remaining iteration is called the "loop remainder" and needs to be accounted for with a final statement after the loop.

From my understanding the original bug (the infinite loop) occurred because this loop remainder was not accounted for and the exit condition was never met.